File: bachetsgame.py
from sys import *
line = stdin.readline()
while line != '':
    s = list(map(int, line.split()))
    n = s[0]
    m = s[1]
    moves = set(s[2:])
    iwin = [False for _ in range(n+1)]
    for i in range(1, n+1):
        for move in moves:
            if i >= move and not iwin[i-move]:
                iwin[i] = True
                break

    if iwin[n]:
        print('Stan wins')
    else:
        print('Ollie wins')

    line = stdin.readline()

File: bachetsgame.cpp
#include <iostream>
using namespace std;

int main() {
    int n, m;
    while(cin >> n >> m){
        int moves[10];
        for(int i=0; i<m; i++){
            cin >> moves[i];
        }

        bool dp[1000001];
        dp[0] = false;
        for(int i=1; i<=n; i++){
            dp[i] = false;
            for(int j=0; j<m; j++){
                if(i>=moves[j] && !dp[i-moves[j]]){
                    dp[i] = true;
                    break;
                }
            }
        }

        if(dp[n]){
            cout << "Stan wins" << endl;
        } else {
            cout << "Ollie wins" << endl;
        }
    }
}